consensus/src/pos/target.rs

```diff
@@ -0,0 +1,567 @@
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use chainstate_types::{BlockIndexHandle, BlockIndexHistoryIterator};
+use common::{
+    chain::{
+        block::ConsensusData, Block, ChainConfig, GenBlock, GenBlockId, PoSChainConfig, PoSStatus,
+        RequiredConsensus,
+    },
+    primitives::{BlockHeight, Compact, Id},
+    Uint256,
+};
+use itertools::Itertools;
+use utils::ensure;
+
+use crate::pos::error::ConsensusPoSError;
+
+fn calculate_average_block_time(
+    chain_config: &ChainConfig,
+    pos_config: &PoSChainConfig,
+    block_id: Id<Block>,
+    block_height: BlockHeight,
+    block_index_handle: &impl BlockIndexHandle,
+) -> Result<u64, ConsensusPoSError> {
+    let history_iter = BlockIndexHistoryIterator::new(block_id.into(), block_index_handle);
+
+    // Determine net upgrade version that current block belongs to.
+    // Across versions config parameters might change so it's invalid to mix versions for target calculations
+    let (_, net_version) = chain_config
+        .net_upgrade()
+        .version_at_height(block_height)
+        .expect("NetUpgrade must've been initialized");
+    let net_version_range = chain_config
+        .net_upgrade()
+        .height_range(net_version)
+        .expect("NetUpgrade must've been initialized");
+
+    // Get timestamps from the history but make sure they belong to the same consensus version.
+    // Then calculate differences of adjacent elements.
+    let block_diffs = history_iter
+        .take(pos_config.block_count_to_average_for_blocktime())
+        .filter(|block_index| net_version_range.contains(&block_index.block_height()))
+        .map(|block_index| block_index.block_timestamp().as_int_seconds())
+        .tuple_windows::<(u64, u64)>()
+        .map(|t| t.0 - t.1);
+
+    let (sum, count) = block_diffs.fold((0u64, 0u64), |(sum, count), curr| (sum + curr, count + 1));
+
+    ensure!(count > 0, ConsensusPoSError::NotEnoughTimestampsToAverage);
+
+    let average = sum / count;
+
+    Ok(average)
+}
+
+fn calculate_new_target(
+    pos_config: &PoSChainConfig,
+    prev_target: &Uint256,
+    average_block_time: u64,
+) -> Result<Compact, ConsensusPoSError> {
+    let average_block_time = Uint256::from_u64(average_block_time);
+    let target_block_time = Uint256::from_u64(pos_config.target_block_time().get());
+
+    // TODO: limiting factor (mintlayer/mintlayer-core#787)
+    let new_target = *prev_target / target_block_time * average_block_time;
```

Comment:
We need to be careful about the operation order here and in `calculate_average_block_time`. This is due to integral division rounding.

When using real numbers, the following give the same result for `a` and `b`:
`a = x / y * z`
`b = x * z / y`

In integer arithmetic, consider `x = 10, y = 11, z = 8`. Then `a = (10 / 11) * 8 = 0 * 8 = 0` whereas `b = (10 * 8) / 11 = 80 / 11 = 7`. For reference, the real valued result is about 7.2727 (= 7 + 3/11) for both expressions. So if `prev_target` is only less than target block time (even by tiny bit), `new_target` will be zero. (Unless the conversion to `Uint256` acts as a fixed point decimal number with extra precision.)

In general, the expression used to calculate `b` is more precise. However, calculating the `x * z` subexpression is more likely to overflow so it may be a good idea to perform it on type of twice the size depending of what sort of range the inputs can be (I see it's `Uint256`, so that should be fine).

We need to be extra careful here also because it's consensus critical code and will be very difficult to change once the thing is running.

Reply:
Ok, if `prev_target` is consistently a couple of orders of magnitude larger a number than `target_block_time`, this is probably not too bad.

Reply:
I will change the order and handle potential overflow